tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ManagedIdentityTests.cs

```diff
@@ -0,0 +1,189 @@
+                .ConfigureAwait(false);
+
+            if (!isEnvironmentVariableSet)
+            {
+                Assert.Fail("AcquireMSITokenForWebAppAsync failed to set environment variables.");
+            }
+
+            //form the URI 
+            string uri = s_baseURL + $"GetMSIToken?azureresource={azureResource}&uri=";
+
+            //Create CCA with Proxy
+            IConfidentialClientApplication cca = CreateCCAWithProxy(uri);
+
+            //Act
+            try
+            {
+                result = await cca.AcquireTokenForClient(s_msi_scopes)
+                    .WithManagedIdentity(userIdentity)
+                    .ExecuteAsync().ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                Trace.WriteLine(ex);
+                Assert.IsNull(ex.Message);
+            }
+
+            //Assert
+            Assert.IsNotNull(result.AccessToken);
+            Assert.AreEqual(TokenSource.IdentityProvider, result.AuthenticationResultMetadata.TokenSource);
+
+            //User Identity Client ID 
+            //Need to check if we are exposing Client ID in msal token response
+            //if (!string.IsNullOrEmpty(userIdentity))
+            //{
+            //    Assert.AreEqual(UserAssignedClientID, result.);
+            //}
+
+            result = await cca.AcquireTokenForClient(s_msi_scopes)
+                .WithManagedIdentity(userIdentity)
+                .ExecuteAsync().ConfigureAwait(false);
+
+            Assert.IsNotNull(result.Scopes);
+            Assert.AreEqual(TokenSource.Cache, result.AuthenticationResultMetadata.TokenSource);
+
+            ClearEnvironmentVariables();
+
+        }
+
+        private async Task<bool> SetEnvironmentVariablesAsync(AzureResource resource)
+        {
+            //Get the Web App Environment Variables from the MSI Helper Service
+            string uri = s_baseURL + "GetEnvironmentVariables?resource=" + resource;
+
+            var environmentVariableResponse = await LabUserHelper
+                .GetMSIEnvironmentVariablesAsync(uri) 
+                .ConfigureAwait(false);
+
+            //process the response
+            if (!string.IsNullOrEmpty(environmentVariableResponse))
+            {
+                s_envVariables = JsonConvert.DeserializeObject<Dictionary<string, string>>(environmentVariableResponse);
+            }
+            else
+            {
+                return false;
+            }
+
+            //Set the environment variables
+            foreach (KeyValuePair<string, string> kvp in s_envVariables)
+            {
+                Environment.SetEnvironmentVariable(kvp.Key, kvp.Value);
+            }
+
+            return true;
+        }
+
+        private void ClearEnvironmentVariables()
+        {
+            //Clear the environment variables
+            foreach (KeyValuePair<string, string> kvp in s_envVariables)
```

Comment:
If env variable A is already present on the box, e.g. A = value1 and you download it from the remote, e.g. A = value2, when you call Clear() you'll reset it. But really what you want to do is to set it back to value1.

Have a look at https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/3849/files#diff-057b5a5346431bf0d06e027a642e673ca6292a88765cc00b0e7603dec26b5154R35 - it captures env variables at point X and on Dispose it resets the state to point X.

Reply:
Fixed all comments except this one, please review the rest. I need to merge main into my branch and work on this, I will hopefully get to this tomorrow. Will update once I am done with it. I also see @neha-bhargava  has merged IMDS, I will also update the tests for IMDS once I get the main